### PR TITLE
[compiled autograd] refactor eager test loading and run custom ops tests

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120679

TestCustomOp's tests uses helper attributes and functions from a util parent class. To support arbitrary test classes, we need to refactor the current approach. Instead of allowlisting certain methods, we can instead copy the whole class and only overwrite the "test_.*" methods.

Compiled autograd fails on ~10/90 of the newly added tests. test_autograd_function_backed_op is the example we discussed in PT-2D meeting about requiring c++ autograd::Function support. I'm addressing this in #120732

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: test/test_custom_ops.py [hunk 1]
```diff
@@ -1842,6 +1843,43 @@ def test_impl_device_invalid(self):
         with self.assertRaisesRegex(RuntimeError, "Expected one of cpu, cuda"):
             torch.library.impl("blah::blah", "somethingsomething")
 
+    def test_autograd_function_backed_op(self):
```
Comment (xmfan) [inline test/test_custom_ops.py:1846]: @zou3519 for custom op test review

File: test/test_custom_ops.py [hunk 2]
```diff
@@ -1842,6 +1843,43 @@ def test_impl_device_invalid(self):
...
+  static torch::autograd::variable_list backward(
+      torch::autograd::AutogradContext *ctx,
+      torch::autograd::variable_list grad_output) {
+    return grad_output;
+  }
+};
+
+torch::Tensor custom_op_with_autograd(torch::Tensor x) {
+  return CustomOpAutogradFunction::apply(x);
+}
```
Comment (zou3519) [inline test/test_custom_ops.py:1866]: This is not exactly what you want: this is a c++ autograd::Function that is bound to Python, not a custom op. To make it a custom op we need some TORCH_LIBRARY macros:
```
TORCH_LIBRARY(mylib, m) {
    m.define(...);
    m.impl(...);
}
```

**Review by jansel — APPROVED** [review]

File: test/test_custom_ops.py (hunk 2 above)
Comment (xmfan) [reply test/test_custom_ops.py:1866]: added the registration, but i'm not sure if the test is invoking the op or the function directly. I get some not declared in scope errors when I try to name the impl function differently than the python one

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**